### ISSUE TITLE
API

### ISSUE THREAD

**Issue by khelle** [tronprotocol/java-tron#78]
Does the JAVA-tron provide any REST / RPC / WebSocket / TCP API or anything like that? If yes, then where can I find the docs? If not, then do I have to use builtin console?

**Comment by Rovak** [comment]
In recent commits it looks like they will be using https://grpc.io/, a start has been made here https://github.com/tronprotocol/java-tron/blob/develop/src/main/java/org/tron/core/services/RpcApiService.java

**Comment by svenanders** [comment]
+1 for the request for documentation. Very interesting to see the direction this project is taking.

**Comment by styk-tv** [comment]
Bits and pieces are coming together from reading the code most of all. You can also refer to original white paper here https://o836fhe91.qnssl.com/tron/whitebook/TronWhitepaper_en.pdf.

I started builds of the develop branch directly into docker container. Pieced a bit from documentation and bit from travis-ci build system into this dockerfile: https://github.com/styk-tv/tron-java-docker/blob/master/Dockerfile This basically tests and builds into a container. Obviously you may need to change settings that are not yet externalized but the idea for me at least was to be able to work with this offline on the train (lots of tunels) So to prevent gradle dependency downloads during run i decided to run all 3 steps inside of container. To run the container just do:

```
polfilm@PetersMacBook2:~$ docker run --name tron -p 10000:10000 -p 50051:50051 -it polfilm/tron-java
Unable to find image 'polfilm/tron-java:latest' locally
latest: Pulling from polfilm/tron-java
af4b0a2388c6: Pull complete
1f7bc6fe3e26: Pull complete
1fa55e432e69: Pull complete
2df1f4f7ff40: Pull complete
c3fd20093ddc: Pull complete
587a45e7c5f4: Pull complete
15b1879141bb: Pull complete
392ccc1ce7a4: Pull complete
Digest: sha256:d5ff5880b8dc183ecbf84a1a1ae4becb1dfd5e3a89530397205d849dce3d865c
Status: Downloaded newer image for polfilm/tron-java:latest
[root@6ad570c4c016 java-tron]# ls -al
total 112
drwxr-xr-x 1 root root 4096 Jan 24 16:38 .
drwxr-xr-x 1 root root 4096 Jan 24 16:37 ..
-rw-r--r-- 1 root root   92 Jan 24 16:31 .codeclimate.yml
drwxr-xr-x 8 root root 4096 Jan 24 16:31 .git
drwxr-xr-x 2 root root 4096 Jan 24 16:31 .github
-rw-r--r-- 1 root root  260 Jan 24 16:31 .gitignore
drwxr-xr-x 1 root root 4096 Jan 24 16:38 .gradle
-rw-r--r-- 1 root root   81 Jan 24 16:31 .travis.yml
-rw-r--r-- 1 root root 1578 Jan 24 16:31 CONTRIBUTING.md
-rw-r--r-- 1 root root 7810 Jan 24 16:31 LICENSE
-rw-r--r-- 1 root root 4337 Jan 24 16:31 README.md
drwxr-xr-x 1 root root 4096 Jan 24 16:38 build
-rw-r--r-- 1 root root 5447 Jan 24 16:31 build.gradle
drwxr-xr-x 2 root root 4096 Jan 24 16:37 database
drwxr-xr-x 3 root root 4096 Jan 24 16:37 database-test
drwxr-xr-x 2 root root 4096 Jan 24 16:31 docker
drwxr-xr-x 3 root root 4096 Jan 24 16:31 gradle
-rwxr-xr-x 1 root root 5296 Jan 24 16:31 gradlew
-rw-r--r-- 1 root root 2260 Jan 24 16:31 gradlew.bat
drwxr-xr-x 4 root root 4096 Jan 24 16:31 infrastructure
drwxr-xr-x 2 root root 4096 Jan 24 16:31 libs
drwxr-xr-x 2 root root 4096 Jan 24 16:37 logs
-rw-r--r-- 1 root root   34 Jan 24 16:31 settings.gradle
drwxr-xr-x 1 root root 4096 Jan 24 16:35 src
[root@6ad570c4c016 java-tron]#
```

At least everything is one place now. So from here I tried several things. Running the jar directly with java had no success. We should probably list independently all possible ways to run the code and with desired configurations. Especially initial node and wallet elements. So far I'm running without Kafka just to see what I can get without it and how far I can get.

I got daemon to run with this command:

```
[root@6ad570c4c016 java-tron]# ./gradlew run -Pserver=true --stacktrace
Starting a Gradle Daemon, 1 incompatible and 1 stopped Daemons could not be reused, use --status for details

> Task :generateProto UP-TO-DATE
Using TaskInputs.file() with something that doesn't resolve to a File object has been deprecated and is scheduled to be removed in Gradle 5.0. Use TaskInputs.files() instead.
        at org.gradle.api.internal.tasks.DefaultTaskInputs$8.validate(DefaultTaskInputs.java:387)
        at org.gradle.api.internal.tasks.StaticValue.validate(StaticValue.java:43)
        at org.gradle.api.internal.tasks.DefaultTaskInputFilePropertySpec.validate(DefaultTaskInputFilePropertySpec.java:124)
        at org.gradle.api.internal.tasks.DefaultTaskInputs.validate(DefaultTaskInputs.java:171)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:47)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:97)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:87)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:248)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:241)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:230)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:123)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:79)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:104)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:98)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:626)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:581)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:98)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
        at java.lang.Thread.run(Thread.java:748)

> Task :compileJava
/home/java-tron/src/main/java/org/tron/common/crypto/ECKey.java:619: error: unmappable character for encoding ASCII
        //        specified in Section 2.3.7, where mlen = ???(log2 p)/8??? or
                                                           ^
/home/java-tron/src/main/java/org/tron/common/crypto/ECKey.java:619: error: unmappable character for encoding ASCII
        //        specified in Section 2.3.7, where mlen = ???(log2 p)/8??? or
                                                            ^
/home/java-tron/src/main/java/org/tron/common/crypto/ECKey.java:619: error: unmappable character for encoding ASCII
        //        specified in Section 2.3.7, where mlen = ???(log2 p)/8??? or
                                                             ^
/home/java-tron/src/main/java/org/tron/common/crypto/ECKey.java:619: error: unmappable character for encoding ASCII
        //        specified in Section 2.3.7, where mlen = ???(log2 p)/8??? or
                                                                        ^
/home/java-tron/src/main/java/org/tron/common/crypto/ECKey.java:619: error: unmappable character for encoding ASCII
        //        specified in Section 2.3.7, where mlen = ???(log2 p)/8??? or
                                                                         ^
/home/java-tron/src/main/java/org/tron/common/crypto/ECKey.java:619: error: unmappable character for encoding ASCII
        //        specified in Section 2.3.7, where mlen = ???(log2 p)/8??? or
                                                                          ^
/home/java-tron/src/main/java/org/tron/common/crypto/ECKey.java:620: error: unmappable character for encoding ASCII
        // mlen = ???m/8???.
                  ^
/home/java-tron/src/main/java/org/tron/common/crypto/ECKey.java:620: error: unmappable character for encoding ASCII
        // mlen = ???m/8???.
                   ^
/home/java-tron/src/main/java/org/tron/common/crypto/ECKey.java:620: error: unmappable character for encoding ASCII
        // mlen = ???m/8???.
                    ^
/home/java-tron/src/main/java/org/tron/common/crypto/ECKey.java:620: error: unmappable character for encoding ASCII
        // mlen = ???m/8???.
                        ^
/home/java-tron/src/main/java/org/tron/common/crypto/ECKey.java:620: error: unmappable character for encoding ASCII
        // mlen = ???m/8???.
                         ^
/home/java-tron/src/main/java/org/tron/common/crypto/ECKey.java:620: error: unmappable character for encoding ASCII
        // mlen = ???m/8???.
                          ^
/home/java-tron/src/main/java/org/tron/common/crypto/ECKey.java:624: error: unmappable character for encoding ASCII
        // conversion routine outputs ???invalid???, then
                                      ^
/home/java-tron/src/main/java/org/tron/common/crypto/ECKey.java:624: error: unmappable character for encoding ASCII
        // conversion routine outputs ???invalid???, then
                                       ^
/home/java-tron/src/main/java/org/tron/common/crypto/ECKey.java:624: error: unmappable character for encoding ASCII
        // conversion routine outputs ???invalid???, then
                                        ^
/home/java-tron/src/main/java/org/tron/common/crypto/ECKey.java:624: error: unmappable character for encoding ASCII
        // conversion routine outputs ???invalid???, then
                                                ^
/home/java-tron/src/main/java/org/tron/common/crypto/ECKey.java:624: error: unmappable character for encoding ASCII
        // conversion routine outputs ???invalid???, then
                                                 ^
/home/java-tron/src/main/java/org/tron/common/crypto/ECKey.java:624: error: unmappable character for encoding ASCII
        // conversion routine outputs ???invalid???, then
                                                  ^
Note: /home/java-tron/src/main/java/org/tron/common/crypto/ECKey.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

> Task :run
23:24:08.641 INFO [NodeImpl] init database
23:24:09.744 INFO [NodeImpl] reset p2p network
cluster: mycluster, uri: udp://127.0.0.1:10000, id: 127.0.0.1:10000
Jan 24, 2018 11:24:11 PM org.tron.core.services.RpcApiService start
INFO: Server started, listening on 50051
<===========--> 85% EXECUTING [47s]
> :run
```

This is as far as it goes for me. Ports that I mapped initially when running docker i got from the last lines above.  UDP at 10K seems to be used for Peer communication (i think).  Port 50051 I'm hoping is the RestApi. Curl returns nothing but then i'm sure this is not a complete setup.

```
polfilm@PetersMacBook2:~$ curl -v http://localhost:50051/account
*   Trying ::1...
* Connected to localhost (::1) port 50051 (#0)
> GET /account HTTP/1.1
> Host: localhost:50051
> User-Agent: curl/7.43.0
> Accept: */*
>
* Empty reply from server
* Connection #0 to host localhost left intact
curl: (52) Empty reply from server
```

So +1 as well for this issue. Now that we have a new wiki at http://wiki.tron.network we should update it with more info as to what to do to get basic setup working. Externalize the settings into config. As I bring this setup into Minikube and then proper Kubernetes with Charts I will be posting updates most likely through wiki PR's  but we all need guidance from project gurus. We know you're busy with releases but let's not loose momentum here with contributors. Another bone would be nice...

**Comment by yangboz** [comment]
I'd  like to fork it, And using Spring-boot+SwaggerUI to wrapper up against the RESTful API.

**Comment by svenanders** [comment]
@styk-tv Tunnel bro ;) Isn't this the same result that you'll get from running the Dockerfile in the main repo? (running with `docker run --name tron -p 10000:10000 -p 50051:50051 -it tron-test`). The RPC server is not implemented yet, so all you'll get are empty replies though.

**Comment by styk-tv** [comment]
@svenanders absolutely, with exception that ALL dependencies get added into container at build hence no waiting on first run. eventually next layer will end up with only binaries required to run.

are you able to run cli wallet and node inside container NOT inteliJ?

**Comment by svenanders** [comment]
@styk-tv Aha, of course. I _just_ noticed that you didn't have to fetch all the deps in your build :)

To run the CLI wallet, you need to setup (or modify) a new gradle task and set the mainClassName to `org.tron.program.example.Tron`). You also need the changes in my PR (literally only two lines in `src/main/java/org/tron/common/application/Application.java`). 

<img width="664" alt="skjermbilde 2018-01-25 kl 22 25 43" src="https://user-images.githubusercontent.com/1560743/35413233-0bd198fc-021f-11e8-9ab8-df2de21bccd0.png">

However, there's still a problem here, because the Blockchain.java doesn't get executed, so the genesis.json is not read and so you won't be able to start with a balance. Thought I'd try and figure out this next.

**Comment by styk-tv** [comment]
@svenanders ok now you're talking my language. thanks for checking. that proves two things: 1. i am not completely insane (output pausing exactly at 85% run) and 2. if we have outside-in approach documented as well it will be easier for everyone to get this setup. my minikube kafka setup fails on communication to zookeeper through service ports. i will document this in new issue and reference here.

lets keep working both of those angles

**Comment by styk-tv** [comment]
@svenanders step forward

1. I used ubuntu xenial container with oracle java 8 instead of tron centos
`docker pull polfilm/tron-java` and lauched with `docker run --name tron -p 10000:10000 -p 50051:50051 -it polfilm/tron-java `
2. i have changed launch to `root@bd2ab4c4f7b6:/home/java-tron# ./gradlew run -PmainClass=org.tron.program.example.Tron`
3. i have updated your 2 line PR 83 suggestion with injection and rebuilt

here is the output I'm getting now
```
> Task :run
16:39:08.353 INFO [NodeImpl] init database
16:39:09.420 INFO [NodeImpl] reset p2p network
cluster: mycluster, uri: udp://127.0.0.1:10000, id: 127.0.0.1:10000
Server localhost: 172.17.0.2
16:39:11.131 INFO [i.a.c.t.n.NettyServer] Binding to /172.17.0.2:5000
16:39:11.243 WARN [i.n.b.ServerBootstrap] Unknown channel option 'TCP_NODELAY' for channel '[id: 0xf38e9ddf]'
16:39:11.309 INFO [i.a.c.t.n.NettyServer] Listening at /172.17.0.2:5000
16:39:11.405 INFO [i.a.c.s.s.ServerContext] /172.17.0.2:5000 - Transitioning to FOLLOWER
16:39:12.549 INFO [i.a.c.s.s.ServerContext] /172.17.0.2:5000 - Transitioning to CANDIDATE
16:39:12.557 INFO [i.a.c.s.s.CandidateState] /172.17.0.2:5000 - Starting election
16:39:12.773 INFO [i.a.c.s.s.ServerContext] /172.17.0.2:5000 - Transitioning to LEADER
16:39:12.801 INFO [i.a.c.s.s.ServerContext] /172.17.0.2:5000 - Found leader /172.17.0.2:5000
16:39:12.977 INFO [i.a.c.s.s.ServerStateMachine] /172.17.0.2:5000 - Taking snapshot 1
16:39:13.139 INFO [i.a.c.s.CopycatServer] Server started successfully!
Server xxd: [ServerMember[type=ACTIVE, status=AVAILABLE, serverAddress=/172.17.0.2:5000, clientAddress=/172.17.0.2:5000]]
Server state: LEADER
16:39:14.006 INFO [BlockChain] new blockchain
16:39:14.361 INFO [i.a.c.t.n.NettyClient] Connecting to /172.17.0.2:5000
16:39:14.451 INFO [i.a.c.t.n.NettyClient] Connected to /172.17.0.2:5000
16:39:14.801 INFO [i.a.c.c.s.ClientSession] Registered session 3
16:39:14.858 INFO [i.a.c.t.n.NettyClient] Connecting to /172.17.0.2:5000
16:39:14.871 INFO [i.a.c.t.n.NettyClient] Connected to /172.17.0.2:5000
16:39:15.270 INFO [Wallet] wallet address: ace9c55a97ebb70e9cf31853000360248fee6540
16:39:15.271 INFO [UTXOSet] reindex
BlockChain loading  ...
Exception in thread "main" java.util.NoSuchElementException: No line found
        at java.util.Scanner.nextLine(Scanner.java:1540)
        at org.tron.common.command.Cli.run(Cli.java:32)
        at org.tron.program.example.Tron.run(Tron.java:65)
        at org.tron.program.example.Tron.main(Tron.java:42)
<===========--> 85% EXECUTING [55s]
> :run

```

still at 85% but more output.

**Comment by Rovak** [comment]
The original question seems to be answered. Feel free to reopen this issue if there are any more related questions
